peregrine/resources/submission/graphql/traversal.py

```diff
@@ -115,35 +115,51 @@ def construct_traversals_from_node(root_node, label_to_subclass):
     return {label: list(paths) for label, paths in traversals.iteritems() if paths}
 
 
-def make_graph_traversal_dict(app_logger):
-    start = time.time()
-    label_to_subclass = {
-        n.__name__: n
-        for n in Node.get_subclasses()
-    }
-    data = {
-        node.label: construct_traversals_from_node(node, label_to_subclass)
-        for node in Node.get_subclasses()
-    }
-    end = int(round(time.time() - start))
-    app_logger.info('Traversed the graph in {} sec'.format(end))
-    return data
+def make_graph_traversal_dict(app, preload=None):
+    """Initialize the graph traversal dict.
+
+    If USE_LAZY_TRAVERSE is False, Peregrine server will preload the full dict at start,
+    or it will be initialized as an empty dict.
+
+    You may call this method with `preload=True` to manually preload the full dict.
+    """
+    app.graph_traversals = getattr(app, 'graph_traversals', {})
+    if preload or not app.config.get('USE_LAZY_TRAVERSE', True):
+        for node in Node.get_subclasses():
+            get_paths_between(node.label, app=app)
+
+
+def get_paths_between(src, dest=None, app=None):
```

Comment:
is there a reason to allow `dest=None`?

Reply:
It's used in the config-controlled backward-compatibility preload at line 129.

Reply:
Oh, I see, thanks. I think for the case where we want to preload let's just call this directly
```python
app.graph_traversals[src] = construct_traversals_from_node(node, label_to_subclass)
```
and have `get_paths_between` always return a value?

Reply:
`app.graph_traversals[src] = ...` feels redundant, but I agree with you to unify return values, please see CRF.